File: 1092-shortest-common-supersequence/1092-shortest-common-supersequence.cpp
class Solution {
public:
    // O(mn) TC & SC
    string shortestCommonSupersequence(string str1, string str2) {
        // we will be using LCS to find SCS
        int m = str1.size(), n = str2.size();
        // 1-based indexing
        vector<vector<int>> lcs(m + 1, vector<int> (n + 1, 0));
        // build the LCS table bottom up using DP
        for(int i = 1; i <= m; ++i) {
            for(int j = 1; j <= n; ++j) {
                // match
                if(str1[i - 1] == str2[j - 1])
                    lcs[i][j] = 1 + lcs[i - 1][j - 1];
                // not a match
                else 
                    lcs[i][j] = max(lcs[i - 1][j], lcs[i][j - 1]);
            }
        }
        // now iterate over the LCS table from the last cell and create a reversed SCS
        string scs = "";
        int row = m, col = n;
        while(row > 0 and col > 0) {
            // if both chars match, we take them only once
            if(str1[row - 1] == str2[col - 1]) {
                scs += str1[row - 1];
                row--; col--;
            } else if(lcs[row - 1][col] > lcs[row][col - 1]) {
                // current value in the LCS table came from the upper row
                // so we take the value from row in the current cell and move up and reduce row
                scs += str1[row - 1];
                row--;
            } else {
                // current value in the LCS table came from the left column
                // so we take the value from column in the current cell and move left and reduce column
                scs += str2[col - 1];
                col--;
            }
        }
        // adding the remaining characters: either in row or column, some characters may be left to be added
        while(row > 0) {
            scs += str1[row - 1];
            row--;
        }
        while(col > 0) {
            scs += str2[col - 1];
            col--;
        }
        // reverse to get the required SCS
        reverse(scs.begin(), scs.end());
        return scs;
    }
};

File: 1092-shortest-common-supersequence/1092-shortest-common-supersequence.java
class Solution {
    public String shortestCommonSupersequence(String str1, String str2) {
        int n1 = str1.length(), n2 = str2.length();
        // Step1: a DP table constructed using LCS
        int[][] dp = new int[n1 + 1][n2 + 1];
        for(int i1 = 1; i1 <= n1; ++i1) {
            for(int i2 = 1; i2 <= n2; ++i2) {
                if(str1.charAt(i1 - 1) == str2.charAt(i2 - 1))
                    dp[i1][i2] = 1 + dp[i1 - 1][i2 - 1];
                else
                    dp[i1][i2] = Math.max(dp[i1 - 1][i2], dp[i1][i2 - 1]);
            }
        }
        
        // Step2: using this table construct the SCS
        int row = n1, col = n2;
        String scs = "";
        while(row > 0 && col > 0) {
            if(str1.charAt(row - 1) == str2.charAt(col - 1)) {
                scs += str1.charAt(row - 1);
                --row;
                --col;
            } else if(dp[row][col - 1] > dp[row - 1][col]) {
                scs += str2.charAt(col - 1);
                --col;
            } else {
                scs += str1.charAt(row - 1);
                --row;
            }
        }
        while(row > 0) {
            scs += str1.charAt(row - 1);
            --row;
        }
        while(col > 0) {
            scs += str2.charAt(col - 1);
            --col;
        }
        return new StringBuilder(scs).reverse().toString();
    }
}